assessment/libs/shapes.php

```diff
@@ -0,0 +1,1465 @@
+      $sidLabLoc[$i] = [$midPt[$i][0]-0.15*($y[$i]-$y[($i+1)%$n])/sqrt(pow($x[$i]-$x[($i+1)%$n],2)+pow($y[$i]-$y[($i+1)%$n],2)),$midPt[$i][1]+0.15*($x[$i]-$x[($i+1)%$n])/sqrt(pow($x[$i]-$x[($i+1)%$n],2)+pow($y[$i]-$y[($i+1)%$n],2))];
+      $args = $args . "text([{$sidLabLoc[$i][0]},{$sidLabLoc[$i][1]}],'".$sidLab[$i]."');";
+    }
+  }
+  
+  if ($hasPoints === true) {
+    $pointLab = array_slice($argsArray[$pointKey],1);
+    for ($i=0;$i<$n;$i++) {
+      $pointLab[$i] = preg_replace('/;/',',',$pointLab[$i]);
+      $args = $args . "dot([$x[$i],$y[$i]]);";
+      $args = $args . "text([$xPointLabPlus[$i],$yPointLabPlus[$i]],'$pointLab[$i]');";
+    }
+  }
+  $gr = showasciisvg("setBorder(10);initPicture(-1.5,1.5,-1.5,1.5);$args;",$size,$size);
+  return $gr;
+}
+
+//--------------------------------------------draw_prismcubes()----------------------------------------------------
+
+// draw_prismcubes("cubes,length,height,depth",["labels,lab1,lab2,lab3"],["size,length"])
+// draw_prismcubes() will draw a cube with no labels.
+// Options include:
+// "cubes,length,height,depth" Draws a rectangular prism made of cubes that is length x height x depth.
+// "labels,lab1,lab2,lab3" This labels the length, height and depth of the prism.
+// "size,length" Sets the sqare image size to be length x length.
+
+function draw_prismcubes() {
+  $size = 300;
+  $input = func_get_args();
+  $argsArray = [];
+  foreach ($input as $list) {
+    if (!is_array($list)) {
+      $list = listtoarray($list);
+    }
+    $list = array_map('trim', $list);
+    $argsArray[]=$list;
+  }
+  foreach ($argsArray as $in) {
+    if ($in[0]=="cubes") {
+      $hasCubes = true;
+    }
+  }
+  if ($hasCubes !== true || empty($argsArray)) {
+    $argsArray[] = ["cubes",1,1,1];
+  }
+
+  foreach ($argsArray as $key => $in) {
+    if ($in[0]=="cubes") {
+      for ($i=1;$i<4;$i++) {
+        if ($in[$i]<1 || abs($in[$i]-round($in[$i]))>1E-9) {
+          echo "Eek! Sides must be whole numbers greater than zero.";
+          return '';
+        }
+      }
+      $length = $in[1];
+      $height = $in[2];
+      $depth = $in[3];
+    }
+    if ($in[0] == "size") {
+      $size = $in[1];
+    }
+    if ($in[0]=="labels") {
+      $labels = array_slice($in,1,3);
+      for ($i=0;$i<3;$i++) {
+        if (!isset($labels[$i])) {
+          $labels[$i] = '';
+        }
+      }
+    }
+  }
+  $xMax = $length+$depth/sqrt(2)+1;
+  $yMax = $height+$depth/sqrt(2)+1;
+  $xyMax = max($xMax,$yMax);
+  $xyLabDiff = $xyMax/12;
+  $xyMin = $xyMax/8;
+  
+  for ($i=0;$i<($height+1);$i++) {
+    $args = $args . "line([0,$i],[$length,$i]);";
+    $args = $args . "line([$length,$i],[$length+$depth/sqrt(2),$i+$depth/sqrt(2)]);";
+  }
+  for ($i=0;$i<($length+1);$i++) {
+    $args = $args . "line([$i,0],[$i,$height]);";
+    $args = $args . "line([$i,$height],[$i+$depth/sqrt(2),$height+$depth/sqrt(2)]);";
+  }
+  for ($i=0;$i<($depth+1);$i++) {
```

Comment:
These three loops seem like they have the potential to generate a LOT of lines, if height, length, or depth was large.  Could we add a max for these?